README.md
# Iconpicker plugin for Craft CMS 

Craft plugin that provides a new field type that offers users an easy way to pick an icon from a .woff or .ttf font file.
You can easily use ionicons or font awesome icons

## Requirements
* Craft 4.0
* PHP 8.0+
* PhenX PHP Font Lib
* One or more .ttf or .woff files

## Example

### Example of the field usage in the template for a cms user
![Screenshot](resources/screenshots/field_example.png)

### Example of the selection popup
![Screenshot](resources/screenshots/popup_example.png)

## Installation
1. Install with Composer
    
       composer require dolphiq/iconpicker
       
2. Install plugin in the Craft Control Panel under Settings > Plugins

3. Add the plugin assets to your application by adding the following line at the begin of your template:
        
        {% do view.registerAssetBundle("plugins\\dolphiq\\iconpicker\\assets\\sharedAsset") %}
        
4. Add the fonts you want to the following directory 
        
        /vendor/dolphiq/iconpicker/src/resources-shared/fonts
               
5. The `Iconpicker Field` type will be available when adding a new field - Settings > Fields > Add new field

## Creating a field with the iconpicker field type
1. Choose the `Iconpicker Field` type
2. When adding a new field with the `Iconpicker Field` type, you can choose which font you want to use for that field from the dropdown

## Using the iconpicker field type
1. Add the field to a field layout (for example to a section)
2. You can now choose an icon when creating or updating a section

## Usage sample in twig template

### Quick example code
        {% do view.registerAssetBundle("plugins\\dolphiq\\iconpicker\\assets\\sharedAsset") %}
        <html style="padding: 0; margin: 0;">
        <head>
            <title></title>

            {{ head() }}

        </head>

        <body>

        {{ beginBody() }}

        {% if entry.iconPickerField.icon %}
            Hex: {{ entry.iconPickerField.icon }}<br>
            {{ entry.iconPickerField.iconSpan|raw }}
        {% endif %}

        {{ endBody() }}

        </body>
        </html>

Display an icon with a custom class:

    <span class='{{ entry.iconClass }} myCustomClass'>{{ entry.iconChar }}</span>


##### Properties of the icon field
1. Get the icon unicode (decimal) 
    
       {{ entry.fieldName.icon }}
    
2. Get the icon unicode (hexadecimal) 

       {{ entry.fieldName.iconHex }}
       
3. Display the icon as html character `&#00000`

       {{ entry.fieldName.iconChar|raw }}
       
4. Display the icon as html character hex `&#xf100` 

       {{ entry.fieldName.iconCharHex|raw }}
       
5. Display the icon as span with character and font class 

       {{ entry.fieldName.iconSpan|raw }}
       
6. Get the icon font class 
       
       {{ entry.fieldName.iconClass }}
       
7. Conditional example of showing icon only when it is set

       {% if entry.fieldName.icon %}
            {{ entry.fieldName.iconSpan|raw }}
       {% endif %}

## Iconpicker Roadmap
- Select and upload the fonts in the UI
- Enable (name) search while picking an icon

### Contributors & Developers
Lucas Weijers - Original developer
Johan Zandstra - Minor changes
Brought to you by [Dolphiq](Https://dolphiq.nl) info@dolphiq.nl


CHANGELOG.md
# Changelog

## Unreleased

### Changed
- Craft 4 compatible.
- Migrate view to Twig template for field input.

### Fixed
- Issues fixed in multi site environments.

## 1.0.5 - 2018-08-07

### Changed
- Added a 'remove icon' button, so you can remove the selected icon
- Fixed a bug that prevented the iconpicker modal from appearing when using the iconpicker field in a matrix field

## 1.0.4 - 2017-12-14

### Changed
- Changed the plugin icon
- Made compatible with Craft RC2
- Use css for the font-asset style set and not scss (sass) to support more environments

## 1.0.3 - 2017-10-06

### Changed
- Removed dynamic resource file from the repository

### Added
- Added screenshots to the readme.
- Added resource links to download font packages

## 1.0.2 - 2017-09-06
- Initial release 


src/fields/Iconpicker.php
<?php
/**
 * Created by PhpStorm.
 * User: lucasweijers
 * Date: 04-09-17
 * Time: 14:58
 *
 * This field will offer the user a choice between various icons in the selected font, and generates the needed css
 */

namespace plugins\dolphiq\iconpicker\fields;

use Craft;
use craft\base\ElementInterface;
use craft\base\Field;
use craft\base\PreviewableFieldInterface;
use craft\helpers\FileHelper;
use FontLib\Font;
use plugins\dolphiq\iconpicker\assets\appAsset;
use plugins\dolphiq\iconpicker\assets\sharedAsset;
use plugins\dolphiq\iconpicker\models\IconpickerModel;
use yii\db\Schema;
use yii\helpers\ArrayHelper;

class Iconpicker extends Field implements PreviewableFieldInterface
{
    // Static
    // =========================================================================

    /**
     * @var string The directory where the fonts are
     */
    const FONT_DIR = '@vendor/dolphiq/iconpicker/src/resources-shared/fonts/';

    /**
     * @var array All extensions that are allowed to be imported as a font
     */
    const FONT_EXT = ['*.woff', '*.ttf'];

    /**
     * @var string Icon class
     */
    const ICON_CLASS = 'dq-icon-';

    /**
     * @var string Pattern to format a safe file name
     */
    const SAFE_NAME_PATTERN = '/[^A-Za-z0-9_\-]/';

    /**
     * @inheritdoc
     */
    public static function displayName(): string
    {
        return Craft::t('app', 'Iconpicker');
    }


    // Properties
    // =========================================================================

    /**
     * @var string|null The input’s placeholder text
     */
    public ?string $placeholder = null;

    /**
     * @var string The type of database column the field should have in the content table
     */
    public string $columnType = Schema::TYPE_STRING;

    /**
     * @var string The current selected iconfont to use
     */
    public string $iconFont = "";

    /**
     * @var array A list with the available fonts
     */
    private array $fonts = [];


    // Public Methods
    // =========================================================================

    /**
     * @inheritdoc
     */
    public function rules(): array
    {
        $rules = parent::rules();
        $rules[] = [['iconFont'], 'safe'];
        return $rules;
    }

    /**
     * @inheritdoc
     */
    public function getSettingsHtml(): ?string
    {
        return Craft::$app->getView()->renderTemplate('dolphiq-iconpicker/settings', [
            'field' => $this,
            'fonts' => $this->getFontOptions(),
        ]);
    }

    /**
     * @inheritdoc
     */
    public function getInputHtml($value, ElementInterface $element = null): string
    {
        // Load the sharedbundle
        $this->getFontCss();

        // Load the assetbundle
        Craft::$app->view->registerAssetBundle(appAsset::class);

        // Display the field
        return Craft::$app->getView()->renderTemplate('dolphiq-iconpicker/input', [
            'name' => $this->handle,
            'value' => $value,
            'field' => $this,
            'icons' => $this->getIcons(),
            'iconFontName' => $this->getIconFontName(),
            'iconClass' => self::ICON_CLASS.$this->getIconFontName(),
        ]);
    }

    /**
     * @inheritdoc
     */
    public function normalizeValue(mixed $value, ElementInterface $element = null): IconpickerModel
    {
        if ($value instanceof IconpickerModel) {
            return $value;
        }

        $model = new IconpickerModel();

        /**
         * Serialised value from the DB
         */
        if (is_string($value)) {
            $value = json_decode($value, true);
        }

        /**
         * Array value from post or unserialised array
         */
        if (is_array($value) && !empty(array_filter($value))) {
            $model->load($value, '');
        }

        return $model;
    }

    /**
     * Generate a css file that creates font families for each font file in the font directory
     *
     * @return void
     * @throws \FontLib\Exception\FontNotFoundException
     * @throws \yii\base\InvalidConfigException
     */
    public function getFontCss(): void
    {
        $sharedAsset = new sharedAsset();
        $scss = "";

        foreach ($this->getFonts() as $safeName => $pathInfo) {
            $fontFile = $pathInfo['path'];
            $font = Font::load($fontFile);
            $font->parse();

            if (!empty($font)) {
                $iconFontName = $safeName;

                if (!empty($iconFontName)) {
                    $scss .= "
@font-face {
    font-family: 'dq-iconpicker-" . $iconFontName . "';
    src: url('../fonts/" . $pathInfo['basename'] . "');
    font-weight: 100;
    font-style: normal;
}\n\n";

                    $scss .= '
[class*="dq-icon-' . $iconFontName . '"] {
  /* use !important to prevent issues with browser extensions that change fonts */
  font-family: dq-iconpicker-' . $iconFontName . ' !important;
  speak: none;
  font-style: normal;
  font-weight: normal;
  font-variant: normal;
  text-transform: none;
  line-height: 1;
  display: inline-block;
  vertical-align: baseline;


  /* Better Font Rendering =========== */
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;

}' . "\n\n";
                }
            }
        }

        file_put_contents(Craft::getAlias($sharedAsset->sourcePath . '/css/fonts.css'), $scss);

        // Register the assetbundle that loads the generated css
        Craft::$app->view->registerAssetBundle(sharedAsset::class);
    }

    /**
     * Get the fontname of the currently selected font.
     *
     * @return string
     */
    public function getIconFontName(): string
    {
        return $this->iconFont;
    }


    // Privtae Methods
    // =========================================================================

    /**
     * Get all the fonts that are residing in the fonts directory and have the right extension.
     * Index them by path as key and pathinfo array as value
     *
     * @return array
     */
    private function getFonts(): array
    {
        if (empty($this->fonts)) {
            $files = FileHelper::findFiles(Craft::getAlias(self::FONT_DIR), ['only' => self::FONT_EXT]);
            $filenames = [];
            $fonts = [];

            foreach ($files as $file) {
                $pathInfo = pathinfo($file);
                $safename = in_array($pathInfo['filename'], $filenames) ? $pathInfo['basename'] : $pathInfo['filename'];
                $safename = $this->safeName($safename);
                $fonts[$safename] = ArrayHelper::merge(
                    [
                        'path' => $file,
                        'safename' => $safename,
                    ],
                    $pathInfo
                );

                $filenames[] = $pathInfo['filename'];
            }

            $this->fonts = $fonts;
        }

        return $this->fonts;
    }

    /**
     * Returns an options list for the settings dropdown when defining a field
     *
     * @return array
     */
    private function getFontOptions(): array
    {
        $f = $this->getFonts();
        if (!empty($f)) {
            return ArrayHelper::map($f, 'safename', 'basename');
        }

        return [];
    }

    /**
     * Load a font and get all unicode characters available in that font.
     *
     * @return array
     * @throws \FontLib\Exception\FontNotFoundException
     */
    private function getIcons(): array
    {
        $returnValue = [];

        if (!empty($this->iconFont)) {
            $fonts = $this->getFonts();
            if (!empty($fonts) && isset($fonts[$this->iconFont])) {
                $font = Font::load($fonts[$this->iconFont]['path']);
                $font->parse();
                $icons = $font->getUnicodeCharMap();
                if ($icons !== null) {
                    foreach ($icons as $dec => $id) {
                        $returnValue[$dec] = dechex($dec);
                    }
                }
            }
        }

        return $returnValue;
    }

    private function safeName($filename): array|string|null
    {
        return preg_replace(self::SAFE_NAME_PATTERN, '-', $filename);
    }
}
